src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2008,6 +2011,115 @@ void CodeGen::genAvxOrAvx2Intrinsic(GenTreeHWIntrinsic* node)
             break;
         }
 
+        case NI_AVX2_GatherVector128:
+        case NI_AVX2_GatherVector256:
+        case NI_AVX2_GatherMaskVector128:
+        case NI_AVX2_GatherMaskVector256:
+        {
+            GenTreeArgList* list = op1->AsArgList();
+            op1                  = list->Current();
+            list                 = list->Rest();
+            op2                  = list->Current();
+            list                 = list->Rest();
+            GenTree* op3         = list->Current();
+            list                 = list->Rest();
+            GenTree* op4         = nullptr;
+            GenTree* lastOp      = nullptr;
+            GenTree* indexOp     = nullptr;
+
+            regNumber op3Reg       = REG_NA;
+            regNumber op4Reg       = REG_NA;
+            regNumber addrBaseReg  = REG_NA;
+            regNumber addrIndexReg = REG_NA;
+            regNumber maskReg      = node->ExtractTempReg(RBM_ALLFLOAT);
+
+            op1Reg = op1->gtRegNum;
+            genConsumeRegs(op1);
+            op2Reg = op2->gtRegNum;
+            genConsumeRegs(op2);
+
+            if (intrinsicId == NI_AVX2_GatherMaskVector128 || intrinsicId == NI_AVX2_GatherMaskVector256)
+            {
+                op4    = list->Current();
+                list   = list->Rest();
+                lastOp = list->Current();
+                op3Reg = op3->gtRegNum;
+                genConsumeRegs(op3);
+                op4Reg = op4->gtRegNum;
+                genConsumeRegs(op4);
+                addrBaseReg  = op2Reg;
+                addrIndexReg = op3Reg;
+                indexOp      = op3;
+
+                // the mask register will be cleared by gather instructions
```

Comment:
I'm not sure I understand the point of this comment. The `maskReg` is a tempreg, so I don't see that it matters whether it will be cleared. Perhaps you mean that we need to copy op4 to a temp reg to use for the mask because it will be over-written?

Reply:
> Perhaps you mean that we need to copy op4 to a temp reg to use for the mask because it will be over-written?

Yes, will update the comment.